Bottom: c845a01f907d303412a00a504c41faea4b8d003d
Top:    aece631eae25ec48c7e0a6d071254d00a8e3e910
Author: Andrew Cooper <andrew.cooper@novatechweb.com>
Date:   2018-07-03 09:12:00 -0500

New role gitlab-sshkey


---
diff --git a/roles/gitlab-sshkey/defaults/main.yml b/roles/gitlab-sshkey/defaults/main.yml
new file mode 100644
index 00000000..4a970d4e
--- /dev/null
+++ b/roles/gitlab-sshkey/defaults/main.yml
@@ -0,0 +1,5 @@
+---
+# defaults file for gitlab-sshkey
+
+gitlab_hostname: "127.0.0.1"
+gitlab_uri: "https://{{ gitlab_hostname }}/"
diff --git a/roles/gitlab-sshkey/tasks/main.yml b/roles/gitlab-sshkey/tasks/main.yml
new file mode 100644
index 00000000..5ef23647
--- /dev/null
+++ b/roles/gitlab-sshkey/tasks/main.yml
@@ -0,0 +1,68 @@
+---
+# tasks file for gitlab-sshkey
+
+- assert:
+    that:
+      - username is defined
+      - pubkey is defined
+      - title is defined
+
+# Login as root user
+- name: Get access token for admin user
+  uri:
+    url: "{{ gitlab_uri }}/oauth/token"
+    method: POST
+    body_format: json
+    body:
+      grant_type: password
+      username: root
+      password: "{{ gitlab_root_password }}"
+    return_content: yes
+  register: auth
+
+# Search for buildbot user
+- name: Find user
+  uri:
+    url: "{{ gitlab_uri }}/api/v4/users?username={{ username }}"
+    method: GET
+    headers:
+      Authorization: "Bearer {{ auth.json.access_token }}"
+    return_content: yes
+  register: founduser
+
+- name: Save new user
+  set_fact:
+    user: '{{ founduser.json[0] }}'
+  when: founduser.json|length > 0
+
+# Get SSH keys for buildbot user
+- name: Get user SSH keys
+  uri:
+    url: "{{ gitlab_uri }}/api/v4/users/{{ user.id }}/keys"
+    headers:
+      Authorization: "Bearer {{ auth.json.access_token }}"
+    return_content: yes
+  register: keys
+
+- name: Delete obsolete keys
+  uri:
+    url: "{{ gitlab_uri }}/api/v4/users/{{ user.id }}/keys/{{ item.id }}"
+    method: DELETE
+    headers:
+      Authorization: "Bearer {{ auth.json.access_token }}"
+    status_code: 204
+  when:
+    - item.title == title or item.key == pubkey
+  loop: "{{ keys.json }}"
+
+- name: Add new key
+  uri:
+    url: "{{ gitlab_uri }}/api/v4/users/{{ user.id }}/keys"
+    method: POST
+    headers:
+      Authorization: "Bearer {{ auth.json.access_token }}"
+    body_format: json
+    body:
+      title: '{{ title }}'
+      key: "{{ pubkey }}"
+    status_code: 201
diff --git a/roles/gitlab-sshkey/vars/main.yml b/roles/gitlab-sshkey/vars/main.yml
new file mode 100644
index 00000000..14b88a19
--- /dev/null
+++ b/roles/gitlab-sshkey/vars/main.yml
@@ -0,0 +1,2 @@
+---
+# vars file for gitlab-sshkey
\ No newline at end of file
